File: src/shapes/sphere.js
import { vec3 } from 'gl-matrix'

export class Sphere {
    #param1;
    #param2;
    #verts;
    #buf;

    constructor(p1, p2) {
        this.#param1 = p1;
        this.#param2 = p2;

        this.#buf = vec3.create();

        this.#verts = []

        const angle = (2 * Math.PI) / Sphere.#param2;
        for(let i = 0; i < param2; i++) {
            this.#makeWedge(i * angle, (i + 1) * angle);
        }
    }

    #makeWedge(currentTheta, nextTheta) {
        const phiStep = (Math.PI / 2) / Sphere.#param1;
        for(let i = 0; i < Sphere.#param1; i++) {
            const nextPhi = i * phiStep;
            const currentPhi = (i + 1) * phiStep;

            const topLeft = vec3.fromValues(
                0.5 * Math.sin(currentPhi) * Math.cos(currentTheta),
                0.5 * Math.cos(currentPhi),
                0.5 * Math.sin(currentTheta) * Math,sin(currentPhi)
            );

            const topRight = vec3.fromValues(
                0.5 * Math.sin(currentPhi) * Math.cos(nextTheta),
                0.5 * Math.cos(currentPhi),
                0.5 * Math.sin(nextTheta) * Math.sin(currentPhi)
            )

            const bottomLeft = vec3.fromValues(
                0.5 * Math.sin(nextPhi) * Math.cos(currentTheta),
                0.5 * Math.cos(nextPhi),
                0.5 * Math.sin(currentTheta) * Math.sin(nextPhi)
            )

            const bottomRight = vec3.fromValues(
                0.5 * Math.sin(nextPhi) * Math.cos(nextTheta),
                0.5 * Math.cos(nextPhi),
                0.5 * Math.sin(nextTheta) * Math.sin(nextPhi)
            )

            this.#makeTile(topLeft, topRight, bottomLeft, bottomRight);
        }
    }

    #makeTile(topLeft, topRight, bottomLeft, bottomRight) {
        this.#insertVec3(topLeft);
        this.#insertVec3(vec3.normalize(this.#buf, topLeft));
        
        this.#insertVec3(bottomLeft);
        this.#insertVec3(vec3.normalize(this.#buf, bottomLeft));

        this.#insertVec3(bottomRight);
        this.#insertVec3(vec3.normalize(this.#buf, bottomRight));

        this.#insertVec3(topLeft);
        this.#insertVec3(vec3.normalize(this.#buf, topLeft));
        
        this.#insertVec3(bottomRight);
        this.#insertVec3(vec3.normalize(this.#buf, bottomRight));

        this.#insertVec3(topRight);
        this.#insertVec3(vec3.normalize(this.#buf, topRight));
    }

    #insertVec3(v) {
        this.#verts.push(v.x, v.y, v.z);
    }
}

File: desktop/src/shapes/Sphere.cpp
#include "Sphere.h"

void Sphere::makeTile(glm::vec3 topLeft,
                      glm::vec3 topRight,
                      glm::vec3 bottomLeft,
                      glm::vec3 bottomRight) {
    // Task 5: Implement the makeTile() function for a Sphere
    // Note: this function is very similar to the makeTile() function for Cube,
    //       but the normals are calculated in a different way!
    insertVec3(m_vertexData, topLeft);
    insertVec3(m_vertexData, glm::normalize(topLeft));

    insertVec3(m_vertexData, bottomLeft);
    insertVec3(m_vertexData, glm::normalize(bottomLeft));

    insertVec3(m_vertexData, bottomRight);
    insertVec3(m_vertexData, glm::normalize(bottomRight));

    insertVec3(m_vertexData, topLeft);
    insertVec3(m_vertexData, glm::normalize(topLeft));

    insertVec3(m_vertexData, bottomRight);
    insertVec3(m_vertexData, glm::normalize(bottomRight));

    insertVec3(m_vertexData, topRight);
    insertVec3(m_vertexData, glm::normalize(topRight));
}

void Sphere::makeWedge(float currentTheta, float nextTheta) {
    // Task 6: create a single wedge of the sphere using the
    //         makeTile() function you implemented in Task 5
    // Note: think about how param 1 comes into play here!
    float phiStep = glm::radians(180.f / (float) m_param1);
    for (int i = 0; i < m_param1; i++) {
        float nextPhi = i * phiStep;
        float currentPhi = (i + 1) * phiStep;

        glm::vec3 topLeft = {
            0.5f * glm::sin(currentPhi) * glm::cos(currentTheta),
            0.5f * glm::cos(currentPhi),
            0.5f * glm::sin(currentTheta) * glm::sin(currentPhi)
        };

        glm::vec3 topRight = {
            0.5f * glm::sin(currentPhi) * glm::cos(nextTheta),
            0.5f * glm::cos(currentPhi),
            0.5f * glm::sin(nextTheta) * glm::sin(currentPhi)
        };

        glm::vec3 bottomLeft = {
            0.5f * glm::sin(nextPhi) * glm::cos(currentTheta),
            0.5f * glm::cos(nextPhi),
            0.5f * glm::sin(currentTheta) * glm::sin(nextPhi)
        };

        glm::vec3 bottomRight = {
            0.5f * glm::sin(nextPhi) * glm::cos(nextTheta),
            0.5f * glm::cos(nextPhi),
            0.5f * glm::sin(nextTheta) * glm::sin(nextPhi)
        };

        makeTile(topLeft, topRight, bottomLeft, bottomRight);
    }
}

void Sphere::setVertexData() {
    float thetaStep = glm::radians(360.f / m_param2);
    for(int i = 0; i < m_param2; i++) {
        float currentTheta = i * thetaStep;
        float nextTheta = (i + 1) * thetaStep;
        makeWedge(currentTheta, nextTheta);
    }
}
